R/textstat_sentiment.R

```diff
@@ -0,0 +1,136 @@
+# textstat_sentiment ----------------
+
+#' Compute sentiment using a dictionary method
+#'
+#' Compute sentiment scores from tokens or document-feature matrices, based on
+#' assigned categories (types or features) of positive, negative, and neutral
+#' sentiment. Several formulas are available, or the user can supply a new one.
+#' @param x a [dfm] or [tokens] object containing sentiment types or features
+#' @param pos,neg character; names of the dictionary keys or features 
+#'   for positive and negative sentiment, respectively
+#' @param neut character; (optional) name of the neutral sentiment feature or 
+#'   key
+#' @param fun function; the formula for computing sentiment, which must refer to
+#'   `pos`, `neg`, and (optionally) `neut`.  Default is the Lowe et al (2001)
+#'   "logit" scale which is the log of (positive / negative) counts.  See
+#'   [sentiment-functions] for details and for additional available functions,
+#'   as well as details on how to supply custom functions.
+#' @param ... additional arguments passed to `fun`
+#' @return a data.frame of sentiment scores
+#' @export
+#' @references  Lowe, W., Benoit, K. R., Mikhaylov, S., & Laver, M. (2011).
+#'   Scaling Policy Preferences from Coded Political Texts. _Legislative Studies
+#'   Quarterly_, 36(1), 123–155.
+#'   <http://doi.org/10.1111/j.1939-9162.2010.00006.x>
+#' @examples
+#' toks <- tail(data_corpus_inaugural, n = 5) %>%
+#'     tokens(remove_punct = TRUE, remove_numbers = TRUE) %>%
+#'     tokens_lookup(dictionary = data_dictionary_LSD2015, nomatch = "_other_")
```

Comment:
here's the `nomatch` argument

Reply:
Why don't you include `tokens_lookup()` this in the new function, so that input objects will be only `toks` and `data_dictionary_xxx`?

Reply:
Actually I started it that way. But we would still need the fields to identify the pos, neg, neutral keys or features.

If the signature were:
```r
textstat_sentiment(x, 
                   dictionary = list(pos = "positive", neg = "negative", neut = NULL),
                   fun = sent_logit)
```
then that would pick up types or features as before, with the option of overwriting the value field, but also work on texts to which dictionaries had not yet been applied.

Reply:
@koheiw Should I implement a version of this and we can see how it looks?

Reply:
How about adding `polarity = c(1, -1)` to dictionaries (as in my comment below), so that sentient analysis can be performed only with tokens and a dictionary?

Reply:
It is just to add a `polarity` or `weight` argument to `dictionary()` and saved in object meta. Users can also save that to using `meta()` in user meta.

Reply:
We can but we also need an argument to override this, since some dictionaries will have multiple possible poles. (LIWC, ANEW, etc). So it still belongs in the argument, but if omitted, we can take it from the dictionary setting.

Reply:
I see. Then, let's make the argument values similar to `c("pos" = 1, "neg" = -1)`. Not only for  consistency but for the flexibility like `c("very_pos" = 2, "pos" = 1, "neg" = -1, "very_neg" = -2)`. The name of the elements should match the dictionary keys of course. The argument default to `NULL`.

Reply:
I think we should leave any weights to the function side, for complete flexibility. See my notes below on the rewrite I just pushed.

Reply:
See ?`sentiment-functions`